### PR TITLE
Combine multiple system messages updates

### LINKED ISSUE #6944: Creating multiple system messages in one request sends multiple requests to the HPB
## Steps to reproduce
1. Create a conversation
2. Setup HPB
3. Add a group with 70 users to the conversation

### Expected behaviour
1 request to the HPB so admin user reloads the chat and sees the system messages

### Actual behaviour
70 requests to the HPB happen

Ref https://github.com/nextcloud/spreed/pull/6943

The problem of origin is that the HPB listens to the `EVENT_AFTER_SYSTEM_MESSAGE_SEND` event.
That should be adjusted to honor `$shouldSkipLastMessageUpdate` and then the new update introduced in #6943 should trigger a final event for the HPB to listen to.

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Only update last_message once, when creating multiple system chat messages
- [x] Combine the update of last message and last activity
- [x] Only push one update to the HPB

Fix #6944

### CODE REVIEW

File: lib/Service/ParticipantService.php [hunk 1]
```diff
@@ -426,9 +433,25 @@ public function addUsers(Room $room, array $participants, ?IUser $addedBy = null
...
+			$lastMessage = $event->getLastMessage();
+			if ($lastMessage instanceof IComment) {
+				$this->updateRoomLastMessage($room, $lastMessage);
+			}
 		}
 	}
 
+	protected function updateRoomLastMessage(Room $room, IComment $message): void {
+		$room->setLastMessage($message);
+		$lastMessageCache = $this->cacheFactory->createDistributed('talk/lastmsgid');
```
Comment (vitormattos) [inline lib/Service/ParticipantService.php:446]: Don't will me more readable?
```suggestion
		$lastMessageCache = $this->cacheFactory->createDistributed('talk/last_message_id');
```

File: lib/Service/ParticipantService.php [hunk 2]
```diff
@@ -426,9 +433,25 @@ public function addUsers(Room $room, array $participants, ?IUser $addedBy = null
...
 			$this->dispatcher->dispatch(Room::EVENT_AFTER_USERS_ADD, $event);
+
+			$lastMessage = $event->getLastMessage();
+			if ($lastMessage instanceof IComment) {
+				$this->updateRoomLastMessage($room, $lastMessage);
+			}
 		}
 	}
 
+	protected function updateRoomLastMessage(Room $room, IComment $message): void {
```
Comment (vitormattos) [inline lib/Service/ParticipantService.php:444]: Is this method only for update cache? If true, maybe will be good expose this at name of method
```suggestion
	protected function updateCacheOfRoomLastMessage(Room $room, IComment $message): void {
```

File: lib/Chat/SystemMessage/Listener.php [hunk 3]
```diff
@@ -408,11 +417,13 @@ protected function sendSystemMessage(Room $room, string $message, array $paramet
 			$referenceId = (string) $referenceId;
 		}
 
-		$this->chatManager->addSystemMessage(
+		return $this->chatManager->addSystemMessage(
```
Comment (vitormattos) [inline lib/Chat/SystemMessage/Listener.php:420]: The method `addSystemMessage` haven't docblock explaining what `shouldSkipLastMessageUpdate` do.
The argument name explain what he does but, what is the impact of skip last message update?
Maybe will be good use docblock in `addSystemMessage` or also in `sendSystemMessage`

The `event->shouldSkipLastMessageUpdate` also haven't explanation about what he does.

File: lib/Service/ParticipantService.php (hunk 1 above)
Comment (nickvergessen) [reply lib/Service/ParticipantService.php:446]: No, this is already in use in other places and we need to purge the correct one

File: lib/Service/ParticipantService.php (hunk 2 above)
Comment (nickvergessen) [reply lib/Service/ParticipantService.php:444]: It's not only about the cache but also about the database

File: lib/Chat/SystemMessage/Listener.php (hunk 3 above)
Comment (nickvergessen) [reply lib/Chat/SystemMessage/Listener.php:420]: > what is the impact of skip last message update?

It skips the update of the last message field in the database

Comment (vitormattos) [reply lib/Chat/SystemMessage/Listener.php:420]: What is the impact when we skip this update? When isn't necessary?

Comment (nickvergessen) [reply lib/Chat/SystemMessage/Listener.php:420]: Added

**Review by vitormattos — APPROVED** [review]

**Pull request merged**